### ISSUE TITLE
What's the relationship between this repo and opendbc

### ISSUE THREAD

**Issue by rusefillc** [HeinrichG-V12/E65_ReverseEngineering#2]
opendbc just got a massive merge of BMW changes https://github.com/commaai/opendbc/commit/a604b15617a1fcad914f10b7f4430dbcf080708a

that's all about https://github.com/commaai/opendbc/blob/master/bmw_e9x_e8x.dbc

Do we see any incompatibilities with E65? Any interest to contribute anything to https://github.com/commaai/opendbc/blob/master/bmw_e9x_e8x.dbc ?

**Comment by rusefillc** [comment]
See also https://github.com/dzid26/BMW_coding/issues/1

**Comment by dzid26** [comment]
Eh. 
It's not that massive.  It's just a lot of corrections to previous findings. 

This repo is amazing though

**Comment by HeinrichG-V12** [comment]
There will be more to come in the days, stay tuned.

I have nothing to do with OpenDBC. :)

This repository contain the definitions from an **E65/E66 especially 760Li and that's how it should stay**! I will not take definitions from other series because I have no way to check them. Here are only definitions that come from official sources (BMW, Bosch, Delphi, ...) or they were checked by me. Here there is only "correct" and no "could be". 

It is very likely that other series have other message definitions which are in conflict with E65/E66 (other DLC, other offset and/or factor, same ID but different content, ...). Pease keep in mind if you take over any CAN definitions, **the K-Matrix is series specific!**

**Comment by dzid26** [comment]
Were you able to obtain checksums calculation formulas from the official sources as well?

Are you planning to create a DBC file for the series? Just wondering.

**Comment by HeinrichG-V12** [comment]
> Were you able to obtain checksums calculation formulas from the official sources as well?

Yes, I will write the formulas for it. It's actually pretty simple:
1. add up all the bytes in the message. The free/unused bits are to be set to 1 in the process. Data area intended for checksum is filled with zeros.
2. add high byte and low byte
3. ignore the overflow and so you have a 8-bit checksum.

I wrote a small C program to test this and it fits. Unfortunately the checksum calculation is message specific because of the unused bits.

> Are you planning to create a DBC file for the series? Just wondering.

It is also planned, but will take time. I don't know yet what I'm going to use for it, Vector CANdb++ or Bosch Busmaster.

**Comment by dzid26** [comment]
You are right. I was [seeing](https://www.spoolstreet.com/threads/coding-m-modules-in-a-non-m-car.6190/post-115073) the similar on E90 generation.
Additionally, usually, the message-id itself is added together with message bytes. 

You can compare it to my python script:
```
def calc_checksum_4bit(work_data, msg_id): # 0x130
  checksum = msg_id
  
  for byte in work_data: #checksum is stripped from the dat
    checksum += byte     #add up all the bytes
  
  checksum = (checksum & 0xFF) + (checksum >> 8); #add upper and lower Bytes
  checksum &= 0xFF #throw away anything in upper Byte
  
  checksum = (checksum & 0xF) + (checksum >> 4); #add first and second nibble
  checksum &= 0xF; #throw away anything in upper nibble

  return checksum

def calc_checksum_8bit(work_data, msg_id): # 0xb8 0x1a0 0x19e 0xaa 0xbf
  checksum = msg_id
  
  for byte in work_data: #checksum is stripped from the data
    checksum += byte     #add up all the bytes
  
  checksum = (checksum & 0xFF) + (checksum >> 8); #add upper and lower Bytes
  checksum &= 0xFF #throw away anything in upper Byte

  return checksum

def calc_checksum_cruise(work_data):# 0x194 checksum is special - initilized with 0 instead of msg_id
  return calc_checksum_8bit(work_data, 0) 
```
Other signals, like accelerometers that were using completely different calculations. Probably crc8 or something. I didn't look deeper. 


As for DBC creation, I was using both Kvaser Database Editor and CANdb++.  Kvaser has a nicer interface, but CANdb++ does more checks on the database so other tools could read with no issues.

**Comment by HeinrichG-V12** [comment]
You are right, the CANID bebelongs to checksum too. This means that CANID and all bytes of the message must be added up.

Do you have a CANID for "other signals"? Maybe there is something in one of the documents.

Since I use CANoe, it will probably be CANdb++. CANdb++ is very restrictive. CANdb++ very often brings errors and warnings where other editors are absolutely quiet.

**Comment by dzid26** [comment]
The ones that I remember that didn't work with normal checksum were 0xD1 and 0xCD. 

Work in progress...:
```
BO_ 209 Accelerometer1: 8 XXX
 SG_ Counter_209 : 52|4@1+ (1,0) [0|255] "" XXX
 SG_ Unknown : 16|16@1- (1,0) [0|65535] "" XXX
 SG_ YawRate : 0|16@1- (1,0) [0|7] "" XXX
 SG_ PitchRate : 32|16@1- (1,0) [0|65535] "" XXX
 SG_ CRC8_209 : 56|8@1+ (1,0) [0|255] "" XXX

BO_ 205 Accelerometer2: 8 XXX
 SG_ Counter_205 : 52|4@1+ (1,0) [0|255] "" XXX
 SG_ LateralAcceleration : 32|16@1- (1,0) [0|255] "" XXX
 SG_ YawRate : 0|16@1- (1,0) [0|255] "" XXX
 SG_ CRC8_205 : 16|16@1- (1,0) [0|65535] "" XXX
```

**Comment by HeinrichG-V12** [comment]
Here is my C code for checksum calculation in 0x130 (terminal status) message:

#define lowByte(w) ((uint8_t) ((w) & 0xff))
#define highByte(w) ((uint8_t) ((w) >> 8))

int main()
{
    uint8_t canId = 0x130;
    uint8_t dlc = 0x5;

    uint16_t temp_w;
    uint8_t cas1_ub, cas1_ub1, cas1_ub2, temp_l, temp_h, temp_ub;

    uint8_t message[5];

    // 0x130 5 45 51 39 B0 48
    message[0] = 0x45;
    message[1] = 0x51;
    message[2] = 0x39;
    message[3] = 0xb0;
    message[4] = (0x48 & 0xf);

    temp_w = canId + message[0] + message[1] + message[2] + message[3] + message[4];

    temp_l = lowByte(temp_w);
    temp_h = highByte(temp_w);

    temp_ub = temp_l + temp_h;

    cas1_ub1 = (temp_ub & 0x0f);
    cas1_ub2 = ((temp_ub >> 4) & 0x0f);
    cas1_ub = cas1_ub1 + cas1_ub2;
    cas1_ub = (cas1_ub & 0xf);

    return 1;
}

And it's looking very good.

I have searched for 0xcd and 0xd1 in the logs from my E65, these messages are not sent in my case. Also in the documentation there are not these messages. Sorry.

**Comment by HeinrichG-V12** [comment]
I just had a quick look into the bmw_e9x_e8x.dbc with CANdb++. Good news: there were no errors when opening :) But, are you absolutely sure about the byte order? I ask because there is Motorola and Intel byte order and I know that BMW and Bosch use mostly Intel byte order. I don't know of any signals that have Motorola byte order.

**Comment by dzid26** [comment]
I didn't know that. The tool I was using was defaulting to Motorola.

But notice signals stored in more than 1 byte have actually Intel order. 
The other ones, due to lack of indication one way or another I didn't bother. 

The only exception in my file I see is BatteryVoltage signal in 0x3B4. This is most likely quite wrong. I didn't focus on it much. It probably should follow your [0x3B4](https://github.com/HeinrichG-V12/E65_ReverseEngineering/blob/main/docs/0x3B4.md). 

Thanks for pointing that out. I need to make it consistent.

**Comment by HeinrichG-V12** [comment]
All signals found here were checked with the logfile of my own E65, so they are correct.
